### PR TITLE
Added support for using a HTTP Proxy when fetching documents from Figma

### PR DESCRIPTION
**Pull request by urpatil**
On AAOS:

Set HTTP Proxy: adb shell settings put global http_proxy "\<host>:\<port>"
Unset: adb shell settings put global http_proxy ":0"

### CODE REVIEW

File: crates/live_update/src/jni.rs [hunk 1]
```diff
@@ -22,20 +22,44 @@ use jni::{JNIEnv, JavaVM};
...
 
 fn throw_basic_exception(env: &mut JNIEnv, msg: String) {
-    //An error occurring while trying to throw an exception shouldn't happen, but let's at least panic with a decent error message
+    // An error occurring while trying to throw an exception shouldn't happen,
+    // but let's at least panic with a decent error message
     env.throw(msg).expect("Error while trying to throw Exception");
 }
 
+fn get_string(env: &mut JNIEnv, obj: &JObject) -> Option<String> {
+    match env.get_string(obj.into()) {
```
Comment (iamralpht) [inline crates/live_update/src/jni.rs:36]: I'm not sure how jni-rs works (we updated from the old one that I knew a bit better ;)), but I'm assuming all of these gets are local refs, so we won't get errors from exhausting global refs etc...

**Review by iamralpht — APPROVED** [review]

File: designcompose/src/androidTest/kotlin/com/android/designcompose/JniLiveWithTokenTests.kt [hunk 2]
```diff
@@ -54,12 +53,12 @@ class JniLiveWithTokenTests {
...
     fun invalidDocId() {
         assertFailsWith<FigmaFileNotFoundException> {
-            LiveUpdateJni.jniFetchDoc("InvalidDocID", firstFetchJson)
+            LiveUpdateJni.jniFetchDoc("InvalidDocID", firstFetchJson, ProxyConfig())
         }
     }
 
     private fun testFetch(docID: String) {
-        with(LiveUpdateJni.fetchDocBytes(docID, firstFetchJson)) {
+        with(LiveUpdateJni.fetchDocBytes(docID, firstFetchJson, ProxyConfig() = null)) {
```
Comment (rylin8) [inline designcompose/src/androidTest/kotlin/com/android/designcompose/JniLiveWithTokenTests.kt:61]: I'm not familiar with this syntax -- what does `ProxyConfig() = null` do?

File: crates/live_update/src/jni.rs [hunk 3]
```diff
@@ -53,7 +77,15 @@ fn jni_fetch_doc<'local>(
         }
     };
 
-    let ser_result = match jni_fetch_doc_impl(&mut env, doc_id, request_json) {
+    let proxy_config: ProxyConfig = match get_proxy_config(&mut env, &jproxy_config) {
+        Ok(it) => it,
+        Err(e) => {
+            eprintln!("Error getting proxy_config - {e}");
```
Comment (timothyfroehlich) [inline crates/live_update/src/jni.rs:83]: Please use the rust logger macros for printing errors

```suggestion
            error!("Error getting proxy_config - {e}");
```

File: designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt [hunk 4]
```diff
@@ -20,14 +20,28 @@ import androidx.annotation.VisibleForTesting
...
+}
+
 internal object LiveUpdateJni {
 
     @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    external fun jniFetchDoc(docId: String, requestJson: String): ByteArray
+    external fun jniFetchDoc(
+        docId: String,
+        requestJson: String,
+        proxyConfig: ProxyConfig
```
Comment (timothyfroehlich) [inline designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt:39]: Why not make this nullable?

File: designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt [hunk 5]
```diff
@@ -20,14 +20,28 @@ import androidx.annotation.VisibleForTesting
 import com.android.designcompose.serdegen.ConvertResponse
 import com.novi.bincode.BincodeDeserializer
 
+// HTTP Proxy configuration.
+internal data class HttpProxyConfig(val proxySpec: String)
+
+// Proxy configuration.
+//
+// Only HTTP proxy supported.
+internal class ProxyConfig {
```
Comment (timothyfroehlich) [inline designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt:29]: How would this be set in a running App?

**Review by timothyfroehlich — CHANGES_REQUESTED** [review]
I've just committed the documentation in #127 and I'd like this to go in with it's usage documented.  Can you rebase this and add documentation for setting the proxy?

**Review by timothyfroehlich — CHANGES_REQUESTED** [review]
I've just committed the documentation in #127 and I'd like this to go in with it's usage documented.  Can you rebase this and add documentation for setting the proxy?

File: designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt (hunk 5 above)
Comment (timothyfroehlich) [reply designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt:29]: Just saw the instructions in the ticket description.

File: designcompose/src/androidTest/kotlin/com/android/designcompose/JniLiveWithTokenTests.kt (hunk 2 above)
Comment (urpatil) [reply designcompose/src/androidTest/kotlin/com/android/designcompose/JniLiveWithTokenTests.kt]: Possibly a paste error.  Fixed.

**Comment by urpatil** [comment]
> I thought when we first learned that some OEMs needed proxy support, there was some issue that made it difficult for us to implement. Did you find a way around that? Was it that we just thought ureq didn't have proper proxy support but it does?

ureq has HTTP proxy support.  The changes are to basically pass the HTTP Proxy setting from Android down to LiveUpdate code in Rust.

File: crates/live_update/src/jni.rs (hunk 3 above)
Comment (urpatil) [reply crates/live_update/src/jni.rs]: Done.

File: designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt (hunk 4 above)
Comment (urpatil) [reply designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt:39]: There was a crash with null JObjects on the Rust side with the get_ methods but not with say a null JString.  Resorted to doing this.

**Comment by urpatil** [comment]
> I've just committed the documentation in #127 and I'd like this to go in with it's usage documented. Can you rebase this and add documentation for setting the proxy?

Done.  Updated live-update/index.md

**Review by rylin8 — APPROVED** [review]

File: designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt (hunk 4 above)
Comment (timothyfroehlich) [reply designcompose/src/main/java/com/android/designcompose/LiveUpdateJni.kt:39]: :/

Ok, probably not that big a deal.

File: docs/_docs/live-update/index.md [hunk 6]
```diff
@@ -71,3 +71,16 @@ App][3].
...
 [3]: {%link _docs/tutorial/index.md %}
+
+## Using an HTTP Proxy Server on Android {#HttpProxy}
+
+Live Update can be configured to download Figma design via an 
+HTTP Proxy Server on Android.
+
+To set the HTTP Proxy Server, run:
+
+```adb shell settings put global http_proxy "<host>:<port>"```
```
Comment (timothyfroehlich) [inline docs/_docs/live-update/index.md:82]: Should set the code style to use and have the block quotes on different lines.
```suggestion
```shell
adb shell settings put global http_proxy "<host>:<port>"
```
```

File: docs/_docs/live-update/index.md [hunk 7]
```diff
@@ -71,3 +71,16 @@ App][3].
...
+
+To unset:
+
+```adb shell settings put global http_proxy ":0"```
```
Comment (timothyfroehlich) [inline docs/_docs/live-update/index.md:86]: This should appear above the list of link references.

**Review by timothyfroehlich — APPROVED** [review]
Otherwise LGTM

File: docs/_docs/live-update/index.md (hunk 6 above)
Comment (urpatil) [reply docs/_docs/live-update/index.md]: Done.

File: docs/_docs/live-update/index.md (hunk 7 above)
Comment (urpatil) [reply docs/_docs/live-update/index.md]: Done.

**Pull request merged**